File: ios/StationToStation/Data/Gossip/GossipPeerPreference.swift
import Foundation

// Lives in GossipBudget.swift on the v2 branch; kept apart while v1's GossipBudget.swift is
// unchanged on main. Folds back in with #462.

/// How long the central would gather sightings before picking one peer, if it had to pick
/// (#444, story 38). **Android holds the same second and a half** (`GOSSIP_PICK_WINDOW_MS`).
///
/// **Provisional.** Nothing has measured it.
let gossipPickWindow: TimeInterval = 1.5

/// The order to try the peers seen in one window, best first (#444, stories 38 and 39).
///
/// Preference, never exclusion. Every candidate is returned — a neighbour with no credit is
/// still pushed to, just later in the list — because a phone that only ever spoke to the
/// neighbours that had already proved useful would never learn that any other one is. The
/// shuffle runs over the whole list before the partition, so both bands are shuffled, and
/// `generator` is a parameter rather than a global for the only reason that matters: a test
/// cannot assert "randomly" without a seed.
///
/// **Not yet wired on this platform**, and that is the honest state rather than an oversight.
/// Android's central holds one connection at a time, so a sighting it takes is a sighting it
/// spends; `GossipTransport` opens a meeting with every peripheral `didDiscover` reports and
/// has no scarce slot to ration. The rule lives here so the two platforms cannot drift apart
/// on what "useful first" means, and so the case has its twin; wiring it is a change to
/// `beginMeeting`, and it only becomes worth making when iOS caps concurrent meetings.
func gossipPreferredPeers(_ candidates: [String], credited: (String) -> Bool,
                          using generator: inout some RandomNumberGenerator) -> [String] {
    var seen = Set<String>()
    let unique = candidates.filter { seen.insert($0).inserted }
    let shuffled = unique.shuffled(using: &generator)
    return shuffled.filter(credited) + shuffled.filter { !credited($0) }
}

File: android/app/src/main/java/io/github/magnusencoded/stationtostation/data/gossip/GossipPeerPreference.kt
package io.github.magnusencoded.stationtostation.data.gossip

// Lives in GossipPolicy.kt on the v2 branch. Kept apart while v1's GossipPolicy.kt has
// different signatures for the functions around it; folds back in with #462.

/**
 * How long the central gathers sightings before it picks one peer to push to (#444, story 38).
 *
 * **Provisional.** The scan reports one device at a time, so without a window there is no set
 * to prefer *within* — the first advertisement seen always wins, and usefulness could never
 * change a decision. A second and a half is long enough to see a second radio in the same
 * room and short enough to stay well inside the time two people stand near each other, which
 * is the same budget the scan mode was chosen against. It costs that much latency on every
 * push; a real night's figures should settle whether that is the right trade.
 */
const val GOSSIP_PICK_WINDOW_MS = 1500L

/**
 * The order to try the peers seen in one window, best first (#444, stories 38 and 39).
 *
 * Preference, never exclusion. Every candidate is returned — a neighbour with no credit is
 * still pushed to, just later in the list — because a phone that only ever spoke to the
 * neighbours that had already proved useful would never learn that any other one is.
 *
 * Ties break randomly, and the shuffle runs over the *whole* list before the partition, so
 * the uncredited band is shuffled too. [random] is a parameter rather than a global for the
 * only reason that matters: a test cannot assert "randomly" without a seed.
 *
 * [credited] answers whether a candidate holds live credit right now; expiry is the caller's
 * business, and `PublicGossipState.prune` has already done it.
 */
fun gossipPreferredPeers(
    candidates: List<String>,
    credited: (String) -> Boolean,
    random: kotlin.random.Random,
): List<String> = candidates.distinct().shuffled(random).sortedByDescending(credited)
